### ISSUE TITLE
Could not send message

### ISSUE THREAD

**Issue by jdneo** [Azure/azure-iot-sdk-node#31]
# Context
- **OS and version used:** Azure Function App
- **Node.js version:** <VERSION> v6.5.0
- **npm version:** <VERSION> 3.10.3
- **list of installed packages:** 

```
+-- azure-iot-common@1.1.6
| `-- crypto@0.0.3
`-- azure-iothub@1.1.8
  +-- azure-iot-amqp-base@1.1.7
  | +-- amqp10@3.5.0
  | | +-- bl@1.2.0
  | | | `-- readable-stream@2.2.6
  | | |   +-- buffer-shims@1.0.0
  | | |   +-- core-util-is@1.0.2
  | | |   +-- inherits@2.0.3
  | | |   +-- isarray@1.0.0
  | | |   +-- process-nextick-args@1.0.7
  | | |   +-- string_decoder@0.10.31
  | | |   `-- util-deprecate@1.0.2
  | | +-- buffer-builder@0.2.0
  | | +-- node-int64@0.4.0
  | | `-- stately.js@1.3.0
  | +-- amqp10-transport-ws@0.0.5
  | | `-- nodejs-websocket@1.7.1
  | +-- bluebird@3.5.0
  | +-- debug@2.6.3
  | | `-- ms@0.7.2
  | `-- uuid@3.0.1
  +-- azure-iot-http-base@1.1.6
  +-- lodash@4.17.4
  `-- uuid@2.0.3
```

- **cloned repo:** 

# Description of the issue:
I'm using azure function app to build a very simple echo service, which means the device send a Ping msg to the iot hub, and it will trigger the function app and the function app will send a C2D Pong msg to the device.

The function app is deployed using script (using azure cli 2.0, Azure REST api and Kudu REST api), after the function app is provisioned, I start to send D2C msg to the IoT Hub, but got 'Could not send message' error. Detailed log is shown below.

And after I stop the device and start it to send D2C msg again, every thing goes ok. The device can receive C2D msg correctly. (nothing else is done)

# Code sample exhibiting the issue:
My function app code: 
```
'use strict';

const Client = require('azure-iothub').Client;
const Message = require('azure-iot-common').Message;
const iotHubConnectionString = process.env['speech-recognition-iothub'];

const cloudClient = Client.fromConnectionString(iotHubConnectionString);
cloudClient.open(function (err) {
    if (err) {
        console.log('Could not connect: ' + err.message);
    } else {
        console.log('Client connected');
    }
});

module.exports = function (context, myEventHubMessage) {
    if (myEventHubMessage.DeviceID) {
        const message = new Message(`Pong - ${new Date().toTimeString()}`);
        cloudClient.send(myEventHubMessage.DeviceID, message, function (err, res) {
            if (err) {
                context.log(`Error in send C2D message: ${err}`);
            } else {
                context.log(`send status: ${res.constructor.name}`);
            }
        });
    }
    context.done();
};
```

# Console log of the issue:
```
2017-04-06T03:53:47  Welcome, you are now connected to log-streaming service.
2017-04-06T03:53:53.341 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.341 Function completed (Success, Id=103c4e2a-cfa0-44af-a7ab-347f87b71999)
2017-04-06T03:53:53.528 Function started (Id=5a0807f7-2f1f-47ce-a437-bf060d91cb92)
2017-04-06T03:53:53.528 Function started (Id=945ebf3f-f513-4c11-bdba-50040f5966ff)
2017-04-06T03:53:53.528 Function started (Id=934a02a9-0dd1-4155-9d85-c0a8fd017c67)
2017-04-06T03:53:53.528 Function started (Id=759801f8-685a-4fb5-91af-e41c23e511dc)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=c44aa3ab-4b23-4d2b-990c-3090d0df7a70)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=6e34c4d1-3d9c-4157-b71c-8b7ef54ee818)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=0da2b730-d2f5-4279-b4a5-938cec4b48b5)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=f387248c-2d09-4d3c-a5d4-b88704ca006d)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=6ee8fe76-a7e5-4b6a-b76d-925c99441254)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=fda4acdb-b040-4afb-8e6f-14adaf82a344)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=b9af8f61-de24-4a2e-8f05-3e9b9bfe8dfc)
2017-04-06T03:53:53.546 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.546 Function started (Id=8324c572-1d95-4681-ba17-2db2d729b51b)
2017-04-06T03:53:53.559 Function started (Id=ec49d0cf-c02d-43c6-8f5f-cbe4df9197c9)
2017-04-06T03:53:53.559 Function started (Id=11c1732d-3eee-41fb-a8b8-3cc6d03ccf61)
2017-04-06T03:53:53.559 Function started (Id=4f4cb112-f711-45d7-a6f9-3006a31c8e11)
2017-04-06T03:53:53.559 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.559 Error in send C2D message: Error: AMQP Transport: Could not send message
2017-04-06T03:53:53.559 Function started (Id=beaefe43-af00-4d53-aee7-feba7704d5c5)
2017-04-06T03:53:53.577 Error in send C2D message: Error: AMQP Transport: Could not send message
```

I've no idea whether it's my code error or issue with Function App, and please let me know if I should provide some other more information.

Thanks!

**Comment by tameraw** [comment]
@jdneo  - Please check the sample available on http://aka.ms/azureiotping
It's supposed to do exactly what you are trying to do.

**Comment by anthonyvercolano** [comment]
I don't see Client connected in the log.

Or are you sure the open has completed when you are trying to send?

**Comment by olivierbloch** [comment]
Azure Functions are stateless, you cannot do the creation and the open on the cloudClient outside of the module.exports function and expect it to still be valid next time the function is triggered. You need to do this in the module.exports function itself as showed in the sample pointed out by @tameraw

**Comment by jdneo** [comment]
Hi, all. Thank you for your reply. I have changed the code according to the azureiotping project. And redeploy a new function app and run the simulatedDevice to send message:

My Azure function code:

```
'use strict';

const Client = require('azure-iothub').Client;
const Message = require('azure-iot-common').Message;
const iotHubConnectionString = process.env['speech-recognition-iothub'];

module.exports = function (context, myEventHubMessage) {
    context.log(myEventHubMessage);
    if (myEventHubMessage.DeviceID) {
        const cloudClient = Client.fromConnectionString(iotHubConnectionString);
        cloudClient.open(function (err) {
            if (err) {
                console.log('Could not connect: ' + err.message);
            } else {
                console.log('Client connected');
                const message = new Message(`Pong - ${new Date().toTimeString()}`);
                cloudClient.send(myEventHubMessage.DeviceID, message, function (err, res) {
                    if (err) {
                        context.log(`Error in send C2D message: ${err}`);
                    } else {
                        context.log(`send status: ${res.constructor.name}`);
                    }
                });
            }
            context.done();
        });
    }
};
```

My simulated device code:

```
'use strict';

var Protocol = require('azure-iot-device-mqtt').Mqtt;
var Client = require('azure-iot-device').Client;
var ConnectionString = require('azure-iot-device').ConnectionString;
var Message = require('azure-iot-device').Message;

var deviceConnectionString = '------';
var deviceId = ConnectionString.parse(deviceConnectionString).DeviceId;

var client = Client.fromConnectionString(deviceConnectionString, Protocol);

var connectCallback = function (err) {
  if (err) {
    console.error('Could not connect: ' + err.message);
  } else {
    consol
```e.log('Client connected');
    client.on('message', function (msg) {
      console.log(`Receive: ${msg.data.toString('utf-8')}`);
      client.complete(msg);
    });

    var sendInterval = setInterval(function () {
      var data = JSON.stringify({
        'DeviceID': deviceId,
        'message': new Date().toTimeString(),
      });
      console.log('Sending:' + data);
      client.sendEvent(new Message(data));
    }, 1000);

    client.on('error', function (err) {
      console.error(err.message);
    });

    client.on('disconnect', function () {
      client.removeAllListeners();
      client.open(connectCallback);
    });
  }
};

client.open(connectCallback);

function printResultFor(op) {
  return function printResult(err, res) {
    if (err) console.log(op + ' error: ' + err.toString());
    if (res) console.log(op + ' status: ' + res.constructor.name);
  };
}

But the error is still there: [full log url](https://shengtemptest.file.core.windows.net/log/iotazurefunction.log?st=2017-04-07T02%3A03%3A00Z&se=2017-04-30T02%3A03%3A00Z&sp=rl&sv=2015-12-11&sr=f&sig=VtfGj5ofMWNEQmXe7QkLVq8BSX00vHFmbGtn%2BDRewWs%3D)

May be it's azure function related issue?

**Comment by jdneo** [comment]
Seems it's because I'm using the dynamic service plan so when I run the npm install, each time the file hiberarchy changed, the azure function will excute once. And reach the quata limit quickly. I'll try to use a dedicate one. And update this issue if I have some new findings. 

Thank you guys!

**Comment by jdneo** [comment]
Hi team. Seems it's not iothub issue. Close this issue for now. Thank you.

**Comment by jdneo** [comment]
Seems I found out the reason. Though I'm not quite sure if it's the real reason.

I have an iot hub for my test work. And I have used it to trigger an azure function. Also I have sent a lot of msgs to it.

And when I was doing the azure provision script. I used the same iot hub to trigger the azure function that provisioned through my script. It seems that the event hub thinks that the msgs I sent before were not consumed. So when the new function app is created, it will be triggered a lot of times (a lot of 'function started show in the console'). And because the function is just created, it has to spend some time to 'wake up'. So it's difficult for the function app to handle such amount of requests. Then a lot of 'Timeout' & 'Could not send message' error appear in the console

If I create a new iot hub, and binding it to a new function app, every thing is fine.
